icaruscode/Decode/DecoderTools/PMTDecoder_tool.cc

```diff
@@ -161,67 +729,422 @@ void PMTDecoder::process_fragment(const artdaq::Fragment &artdaqFragment)
-        std::cout << "      size: " << ev_size_quad_bytes << ", data size: " << data_size_double_bytes << ", samples/channel: " << nSamplesPerChannel << ", time: " << time_tag << std::endl;
+        mf::LogVerbatim(fLogCategory)
+          << "----> PMT Fragment ID: " << std::hex << fragment_id << std::dec
+            << ", boardID: " << boardId
+            << ", nChannelsPerBoard: " << nChannelsPerBoard
+            << ", nSamplesPerChannel: " << nSamplesPerChannel
+            << ", enabled (" << nEnabledChannels << "): " << icarus::ns::util::bin(enabledChannels)
+          << "\n      size: " << ev_size_quad_bytes
+            << ", data size: " << data_size_double_bytes
+            << ", samples/channel: " << nSamplesPerChannel
+            << ", trigger time tag: " << time_tag
+          ;
     }
 
     const uint16_t* data_begin = reinterpret_cast<const uint16_t*>(artdaqFragment.dataBeginBytes() + sizeof(sbndaq::CAENV1730EventHeader));
-    const uint16_t* value_ptr  =  data_begin;
-    uint16_t        value      = 0;
-    size_t          ch_offset  = 0;
-
-    // Temporary? 
-    time_tag = 0;
 
     // Recover the information for this fragment
-    if (fChannelMap->hasPMTDigitizerID(fragment_id))
+    if (fChannelMap.hasPMTDigitizerID(eff_fragment_id))
     {
-        const icarusDB::DigitizerChannelChannelIDPairVec& digitizerChannelVec = fChannelMap->getChannelIDPairVec(fragment_id);
+        assert(fBoardInfoLookup);
+    
+        /*
+         * The trigger time is always the nominal one, because that is the
+         * reference time of the whole DAQ (PMT, TPC...).
+         * We only need to know how sooner than the trigger the V1730 buffer
+         * starts. Oh, and the delay from the global trigger time to when
+         * the readout board receives and processes the trigger signal.
+         */
+        details::BoardInfoLookup::BoardInfo_t const* boardInfo = fBoardInfoLookup->findBoardInfo(fragment_id);
+        if (!boardInfo) {
+            if (fRequireKnownBoards) {
+                cet::exception e("PMTDecoder");
+                e << "Input fragment has ID " << fragment_id
+                  << " which has no associated board information (`BoardSetup`";
+                if (!hasPMTconfiguration()) e << " + `.FragmentID`";
+                throw e << ").\n";
+            }
+        }
+        else {
+            assert(boardInfo->fragmentID == fragment_id);
+            assert(boardInfo->setup);
+        }
+        
+        NeededBoardInfo_t const info = fetchNeededBoardInfo(boardInfo, fragment_id);
+        
+        nanoseconds const preTriggerTime = info.preTriggerTime;
+        nanoseconds const PMTtriggerDelay = info.PMTtriggerDelay;
+        auto const timeStamp
+          = fDetTimings.TriggerTime() - PMTtriggerDelay - preTriggerTime;
```

Comment:
Just a comment here. In case you want to start preparing for the multi-acquisition readout you may start playing with the Withe Rabbit time from the Trigger Fragment class and the timeStampNSec from the Fragment overlay class. For legacy purposes it should be easy to apply a fix: the TTT doesn't mean anything if the reset is not sent with the PPS, but I think the timeStampSec method has the NPT server time stored regardless, so you'll lose nanosecond precision but still have something "meaningful" that appears and should be comparable with the Withe Rabbit time. Hope it makes sense.

Reply:
What do you mean with "legacy purposes" here?
For the rest, the only thing I need is a way to find out which waveform matches the global trigger.
I am not sure if the fragment timestamp (which I think is what you refer to with `timeStampSec`) helps there - I'll ask DAQ people.